### PR TITLE
[PIBD] PMMR Desegmenter Structure (Pt. 1)

### PR DESCRIPTION
**Pull request by yeastplume**
This is the first part of several anticipated PRs to enable and test the reconstruction of the txhashset from PMMR segments. Note that none of the code introduced here is called from anywhere but the main test `test_pibd_copy`. Merging this a bit earlier than I'd planned, but breaking up the work like this keeps the review process a bit more manageable, and the included test is needed to help track down a potential serious performance issue in header sync 😄 

Included in this PR:

* Introduce a `desegmenter` type containing functions to validate and append PMMR segments.
* Add optional instance of desegmenter to chain, with lazy init method
* Function in `BitmapAccumulator` to return the accumulator as a raw bitmap for use in desegmenter validation
* Add functions in txhashset extension to apply outputs, rangeproofs and kernels within an extension
* Add test to copy sample dataset from previous PR #3665 into new txhashset.
* Fix for SegmentIdentifier returning wrong number of segments in `count_segments_required`
* change `last_pos` to `mmr_size` in Segment to be consistent with changes #3666 

Note the current copy tests do work, however they haven't been tested or verified against actual pruned data.

### CODE REVIEW

File: chain/src/chain.rs [hunk 1]
```diff
@@ -863,6 +865,43 @@ impl Chain {
 		))
 	}
 
+	/// instantiate desegmenter (in same lazy fashion as segmenter, though this should be be
```
Comment (tromp) [inline chain/src/chain.rs:868]: to be be or not to be be...

File: chain/src/txhashset/desegmenter.rs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+		)?;
+		Ok(())
+	}
+
+	// Calculate and store number of leaves and positions in the bitmap mmr given the number of
+	// outputs specified in the header. Should be called whenever the header changes
+	fn calc_bitmap_mmr_sizes(&mut self) {
+		// Number of leaves (BitmapChunks)
+		self.bitmap_mmr_leaf_count =
+			(pmmr::n_leaves(self.archive_header.output_mmr_size) as f64 / 1024f64).ceil() as u64;
```
Comment (tromp) [inline chain/src/txhashset/desegmenter.rs:126]: Don't use floating point!
(n+1023) / 1024 computes rounded up division with plain integers.

File: chain/src/txhashset/desegmenter.rs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+			.unwrap_or(&pmmr::insertion_to_pmmr_index(self.bitmap_mmr_leaf_count))
+			.clone();
+		debug!(
+			"pibd_desgmenter - expected size of bitmap MMR: {}",
+			self.bitmap_mmr_size
+		);
+	}
+
+	/// Adds and validates a bitmap chunk
+	/// TODO: Still experimenting, this expects chunks received to be in order
```
Comment (tromp) [inline chain/src/txhashset/desegmenter.rs:143]: Why do you need them in order?

File: core/src/core/pmmr/segment.rs [hunk 4]
```diff
@@ -87,7 +87,8 @@ impl SegmentIdentifier {
 	/// Returns number of segments required that would needed in order to read a
 	/// pmmr of size `target_mmr_size` in segments of height `segment_height`
 	pub fn count_segments_required(target_mmr_size: u64, segment_height: u8) -> usize {
-		pmmr::n_leaves(target_mmr_size) as usize / (1 << segment_height as usize)
+		(pmmr::n_leaves(target_mmr_size) as f64 / (1 << segment_height as usize) as f64).ceil()
```
Comment (tromp) [inline core/src/core/pmmr/segment.rs:90]: No floating point please! See other comment.

File: chain/src/txhashset/txhashset.rs [hunk 5]
```diff
@@ -1232,6 +1239,33 @@ impl<'a> Extension<'a> {
...
+			self.output_pmmr
+				.push(&output_identifier)
+				.map_err(&ErrorKind::TxHashSetErr)?;
+		}
+		Ok(())
+	}
+
+	/// Apply a rangeproof segment to the output PMMR. must be called in order
+	/// TODO: Not complete
+	pub fn apply_rangeproof_segment(&mut self, segment: Segment<RangeProof>) -> Result<(), Error> {
```
Comment (tromp) [inline chain/src/txhashset/txhashset.rs:1259]: why feed in a whole segment when you only use one component?

File: chain/src/txhashset/txhashset.rs [hunk 6]
```diff
@@ -1232,6 +1239,33 @@ impl<'a> Extension<'a> {
...
+		let (_sid, _hash_pos, _hashes, _leaf_pos, leaf_data, _proof) = segment.parts();
+		for proof in leaf_data {
```
Comment (tromp) [inline chain/src/txhashset/txhashset.rs:1261]: too much copy-paste?!
shouldn't you use the proof component here?

File: chain/tests/test_pibd_copy.rs [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+	clean_output_dir(&dest_root_dir);
+}
+
+#[test]
+#[ignore]
+// As above, but run on a real instance of a chain pointed where you like
+fn test_pibd_copy_real() {
+	util::init_test_logger();
+	// if testing against a real chain, insert location here
+	let src_root_dir = format!("/Users/yeastplume/Projects/grin_project/server/chain_data");
```
Comment (tromp) [inline chain/tests/test_pibd_copy.rs:228]: This is not gonna work on my machine:-(

File: chain/tests/test_pibd_validation.rs [hunk 8]
```diff
@@ -87,7 +87,7 @@ fn test_pibd_chain_validation_impl(is_test_chain: bool, src_root_dir: &str) {
 		println!("BITMAP PMMR NUM_LEAVES: {}", bitmap_mmr_num_leaves);
 
 		// And total size of the bitmap PMMR
-		let bitmap_pmmr_size = pmmr::peaks(bitmap_mmr_num_leaves + 1)
+		let bitmap_pmmr_size = pmmr::peaks(bitmap_mmr_num_leaves)
```
Comment (tromp) [inline chain/tests/test_pibd_validation.rs:90]: I assume this is because the former was written before my 1-based -> 0-based conversion?

**Review by tromp — APPROVED** [review]
Not a particularly through review, but found some minor points to be addressed...

**Comment by yeastplume** [comment]
Thank you, will address bits before merging this. It's perhaps a bit early for review and there are still a lot of things here that are going to change, but I'd just like to get everything I have to date merged before continuing with this.

File: chain/tests/test_pibd_copy.rs (hunk 7 above)
Comment (yeastplume) [reply chain/tests/test_pibd_copy.rs:228]: it's not meant to, it's meant for running a manual test against a live chain instance (so while testing you modify the path).. I'll make this more clear in comments

File: core/src/core/pmmr/segment.rs (hunk 4 above)
Comment (yeastplume) [reply core/src/core/pmmr/segment.rs:90]: 👍

File: chain/src/txhashset/desegmenter.rs (hunk 2 above)
Comment (yeastplume) [reply chain/src/txhashset/desegmenter.rs:126]: :+1:

File: chain/tests/test_pibd_validation.rs (hunk 8 above)
Comment (yeastplume) [reply chain/tests/test_pibd_validation.rs:90]: yes

**Comment by yeastplume** [comment]
Addressed a few issues, will keep the rest in consideration during future work

**Pull request merged**